cmd/boulder-migrator/main.go

```diff
@@ -0,0 +1,329 @@
+	if err != nil {
+		tx.Rollback()
+		return
+	}
+	err = tx.Commit()
+	return
+}
+
+// list applied migrations
+func listMigrations(Sql *sql.DB) (err error) {
+	rows, err := Sql.Query("SELECT id, description, applied FROM migrations")
+	if err != nil {
+		return
+	}
+	defer rows.Close()
+	fmt.Println("# Currently applied migrations\n")
+	for rows.Next() {
+		var id int
+		var desc string
+		var appliedInt interface{}
+		var applied time.Time
+		if err = rows.Scan(&id, &desc, &appliedInt); err != nil {
+			return
+		}
+		applied, ok := appliedInt.(time.Time)
+		if !ok {
+			fmt.Println(ok, applied)
+			return
+		}
+		fmt.Printf("%d: [%s]\n\t%s\n", id, applied, desc)
+	}
+	return
+}
+
+// apply new migration
+func applyMigration(Sql *sql.DB, migration Migration, yes bool) (err error) {
+	fmt.Printf("[migration %d]\n", migration.Id)
+	fmt.Printf("\tDescription: %s\n", migration.Desc)
+	fmt.Printf("\n\tMigrate SQL:\n")
+	for _, sqlLine := range migration.UpSQL {
+		fmt.Printf("\t%s\n", sqlLine)
+	}
+	fmt.Printf("\n\tRevert SQL:\n")
+	for _, sqlLine := range migration.DownSQL {
+		fmt.Printf("\t%s\n", sqlLine)
+	}
+	if !yes {
+		if answer := getYN("\nWould you like to apply this migration"); !answer {
+			fmt.Println("Ok, bye!")
+			return
+		}
+	}
+	tx, err := Sql.Begin()
+	if err != nil {
+		return
+	}
+	for _, sqlStmt := range migration.UpSQL {
+		_, err = tx.Exec(sqlStmt)
+		if err != nil {
+			tx.Rollback()
+			return
+		}
+	}
+	for _, sqlStmt := range migration.DownSQL {
+		sqlStmt = strings.TrimRight(sqlStmt, ";")
+	}
+	_, err = tx.Exec("INSERT INTO migrations (id, description, applied, down) VALUES (?, ?, ?, ?)", migration.Id, migration.Desc, time.Now(), strings.Join(migration.DownSQL, ";"))
+	if err != nil {
+		tx.Rollback()
+		return
+	}
+	err = tx.Commit()
+	if err != nil {
+		return
+	}
+	fmt.Printf("Migration %d applied\n", migration.Id)
+	return
+}
+
+// revert last migration
+func revertMigration(Sql *sql.DB) (err error) {
+	var id      int
+	var downSql string
+	err = Sql.QueryRow("SELECT id, down FROM migrations ORDER BY id DESC LIMIT 1").Scan(&id, &downSql)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			fmt.Println("No migrations to revert.")
+			err = nil
+		}
+		return
+	}
+	tx, err := Sql.Begin()
+	if err != nil {
+		return
+	}
+	for _, sqlStmt := range strings.Split(downSql, ";") {
+		_, err = tx.Exec(sqlStmt)
+		if err != nil {
+			tx.Rollback()
+			return
+		}
+	}
+	_, err = tx.Exec("DELETE FROM migrations WHERE id = ?", id)
+	if err != nil {
+		tx.Rollback()
+		return
+	}
+	fmt.Printf("Migration %d reverted\n", id)
+	err = tx.Commit()
+	return
+}
+
+// revert to specific migration
+func revertMigrationTo(Sql *sql.DB, id int, yes bool) (err error) {
+	var revertCount int
+	err = Sql.QueryRow("SELECT count(*) FROM migrations WHERE id > ?", id).Scan(&revertCount)
+	if err != nil {
+		return
+	}
+	if revertCount == 0 {
+		fmt.Printf("Migration [ID: %d] does not exist\n", id)
```

Comment:
This error is not quite right: It should be more like "There are no migrations after ID %d to revert."